gitian-linux: Build binaries for 64-bit POWER

luke-jr opened the pull request:
A set of both big and little endian binaries, should be compatible with PowerPC 970 (Apple G5) and newer.

~~Also splits libpng out of Qt (since Qt's bundled copy is broken on POWER) and disables JPEG (since we don't use it).~~

Tested only the little endian variant, on Gentoo.

~~(Based on #14065)~~

luke-jr committed:
Support glibc-back-compat on 64-bit POWER

luke-jr committed:
gitian-linux: Build binaries for 64-bit POWER

luke-jr committed:
symbol-check: Fix readelf output parsing

luke-jr committed:
gitian: Properly quote arguments in wrappers

luke-jr commented:
Also splits libpng out of Qt (since Qt's bundled copy is broken on POWER) and disables JPEG (since we don't use it).

laanwj commented:
Concept ACK

awilfox commented:
At least RHEL, SuSE, and Adélie ship big-endian images.

Big endian is simple to run in KVM paravirt on the little-endian systems with a performance hit of less than 1%.  This should be very easy to test, especially with Gentoo, since Gentoo can make a cross-root somewhere and then you can just point KVM at that disk.

Any POWER7 or older system, including most workstations before the Talos (including IBM's post-RS and Apple's 970), are exclusively big endian.

Any system (of any generation) running openSuSE Tumbleweed or Adélie is exclusively big endian.

laanwj commented:
> Any POWER7 or older system, including most workstations before the Talos (including IBM's post-RS and Apple's 970), are exclusively big endian.

Those are outside the scope of this PR, though, OP mentions this is POWER8+.

luke-jr commented:
Rebased

theuni commented on contrib/gitian-descriptors/gitian-linux.yml:
```diff
@@ -79,7 +85,11 @@ script: |
             echo "export FAKETIME=\"$1\"" >> ${WRAP_DIR}/${i}-${prog}
-            echo "exec \"\$REAL\" \"\$@\"" >> $WRAP_DIR/${i}-${prog}
+            if [ "${i:0:9}" = "powerpc64" ]; then
+                echo "exec \"\$REAL\" -mcpu=power8 -mtune=power9 \"\$@\"" >> $WRAP_DIR/${i}-${prog}
```
This looks like something that should be handled in depends/hosts/linux.mk

luke-jr replied:
What if someone else wants to build for POWER7 (outside of gitian)?

awilfox replied:
Is there a reason for specifying `-mcpu=power8` here at all?  VSX goes back to P7 and VMX goes back to P4…

luke-jr replied:
Maybe not. How far back can we safely go without a noticeable performance hit?

Would also be good to confirm #13203 still works if compiled for pre-POWER8.

awilfox replied:
`vec_vsx_ld` is POWER7; but the SHA sigma instructions are exclusive to POWER8 and newer.  However, it looks like the merged code does check the hardware cap before using it, so it depends on compiler support, not build hardware support.  (That is, if my understanding is correct, one could build all the way down to -mcpu=generic and you'd still get speed on P8+.)

theuni commented on src/Makefile.am:
```diff
@@ -573,7 +573,7 @@ clean-local:
 if GLIBC_BACK_COMPAT
 	@echo "Checking glibc back compat..."
-	$(AM_V_at) READELF=$(READELF) CPPFILT=$(CPPFILT) $(top_srcdir)/contrib/devtools/symbol-check.py < $(bin_PROGRAMS)
+	$(AM_V_at) READELF=$(READELF) CPPFILT=$(CPPFILT) $(top_srcdir)/contrib/devtools/symbol-check.py $(BITCOIN_SYMBOL_CHECK_OPTIONS) < $(bin_PROGRAMS)
```
What is BITCOIN_SYMBOL_CHECK_OPTIONS ? I don't see any references to it, is this just intended to be an optional env var for local testing?

theuni reviewed (commented):
Concept ACK, a few questions though.

laanwj commented on contrib/devtools/symbol-check.py:
```diff
@@ -111,12 +111,13 @@ def read_symbols(executable, imports=True):
+        words = line.split()
+        if 'Machine:' in words:
+            arch = words[-1]
+        m = re.match(r'^\s*\d+:\s*[\da-f]+\s+\d+\s(?:(?:\S+\s+){3})(?:\[.*\]\s+)?(\S+)\s+(\S+).*$', line)
```
ahhh
even parsing the ELF directly would be easier here

luke-jr replied:
This is a fairly straightforward regex... Do we really need to ban regexs? :/

Looking at alternatives:

* Debian's linter uses almost the same regex (they tolerate multiple spaces in one place this didn't).
* Perl's Parse::Readelf module does some fancy parsing of column headers and stuff, and doesn't really seem like it works (not to mention only supports *printing* its output, not actual code usage), and adds a lot of complexity we really don't need.
* pyelftools doesn't seem to get symbol version information for some reason.

luke-jr commented on configure.ac:
```diff
@@ -674,6 +674,11 @@ if test x$use_glibc_compat != xno; then
   AX_CHECK_LINK_FLAG([[-Wl,--wrap=log2f]], [COMPAT_LDFLAGS="$COMPAT_LDFLAGS -Wl,--wrap=log2f"])
+  case $host in
+    powerpc64* | ppc64*)
+      AX_CHECK_LINK_FLAG([[-Wl,--no-tls-get-addr-optimize]], [COMPAT_LDFLAGS="$COMPAT_LDFLAGS -Wl,--no-tls-get-addr-optimize"])
```
It *should* work without the no- prefix, but upon further research, I'm not sure if this is the correct solution at all.

According to [the documentation](https://sourceware.org/binutils/docs/ld/PowerPC64-ELF64.html), systems without the glibc-side part can still explicitly enable the option and produce binaries that work (with a performance hit) on older systems. I'm not sure how this is technically implemented.

luke-jr replied:
Managed to get an old glibc running and confirmed `--no-tls-get-addr-optimize` is in fact needed. No clue why the docs imply otherwise. :/

luke-jr replied:
As for the original comment:

1. At least GNU ld errors if --no-foo is provided.
2. Other platforms might not have compatibility issues using this flag.

luke-jr commented on depends/packages/libpng.mk:
```diff
@@ -0,0 +1,23 @@
+endef
+
+define $(package)_build_cmds
+  $(MAKE) $($(package)_build_opts) PNG_COPTS='-fPIC'
```
No idea why, but it doesn't seem to work the usual way via configure.

```
/usr/lib/gcc-cross/riscv64-linux-gnu/8/../../../../riscv64-linux-gnu/bin/ld: /home/ubuntu/build/bitcoin/depends/riscv64-linux-gnu/lib/libpng16.a(png.o): relocation R_RISCV_HI20 against `a local symbol' can not be used when making a shared object; recompile with -fPIC
/usr/lib/gcc-cross/riscv64-linux-gnu/8/../../../../riscv64-linux-gnu/bin/ld: /home/ubuntu/build/bitcoin/depends/riscv64-linux-gnu/lib/libpng16.a(pngerror.o): relocation R_RISCV_HI20 against `a local symbol' can not be used when making a shared object; recompile with -fPIC
/usr/lib/gcc-cross/riscv64-linux-gnu/8/../../../../riscv64-linux-gnu/bin/ld: /home/ubuntu/build/bitcoin/depends/riscv64-linux-gnu/lib/libpng16.a(pngget.o): relocation R_RISCV_HI20 against `__stack_chk_guard@@GLIBC_2.27' can not be used when making a shared object; recompile with -fPIC
...
```

practicalswift commented:
Concept ACK

practicalswift commented:
@luke-jr What setup would you suggest for reviewers who want to test those binaries?

awilfox commented:
You could spin up an IntegriCloud VPS, but I don't know of any big endian distros that use glibc.  All of them that I am aware of (Adelie, Void, etc) went to musl since it is easier to support BE on musl.

luke-jr commented:
Rebased, and extended support back to the PowerPC 970 (Apple G5) since it didn't seem to have a notable performance hit on POWER9 (it actually performed better!).

practicalswift commented:
@luke-jr What setup would you suggest for reviewers who want to test those binaries?

luke-jr commented:
@awilfox answered that above. If you prefer to buy real hardware, I'm using https://www.raptorcs.com/content/base/products.html

MarcoFalke commented:
Fails with:

```
> make[1]: Leaving directory '/home/ubuntu/build/bitcoin/depends/work/build/powerpc64-linux-gnu/qt/5.9.7-72625d07ee6/qtbase/config.tests/xcb_xlib'
 => source accepted.
test config.gui.libraries.xcb_xlib succeeded
Checking for XKB config root... 
test config.gui.tests.xkbconfigroot gave result /usr/share/X11/xkb
Checking for xkbcommon... 
Trying source 0 (type pkgConfig) of library xkbcommon ...
+ PKG_CONFIG_SYSROOT_DIR=/ PKG_CONFIG_LIBDIR=/home/ubuntu/build/bitcoin/depends/powerpc64-linux-gnu/lib/pkgconfig /usr/bin/pkg-config --exists --silence-errors xkbcommon
pkg-config did not find package.
  => source produced no result.
test config.gui.libraries.xkbcommon FAILED
Done running configuration tests.

Configure summary:

Building on: linux-g++ (x86_64, CPU features: mmx sse sse2)
Building for: bitcoin-linux-g++ (power64, CPU features: altivec)
Configuration: cross_compile enable_new_dtags largefile precompile_header silent release c++11 dbus no-qml-debug reduce_exports release_tools static
Build options:
  Mode ................................... release; optimized tools
  Optimize release build for size ........ no
  Building shared libraries .............. no
  Using C++ standard ..................... C++11
  Using ccache ........................... no
  Using gold linker ...................... no
  Using new DTAGS ........................ yes
  Using precompiled headers .............. yes
  Using LTCG ............................. no
  Target compiler supports:
  Build parts ............................ libs
Qt modules and options:
  Qt Concurrent .......................... no
  Qt D-Bus ............................... yes
  Qt D-Bus directly linked to libdbus .... no
  Qt Gui ................................. yes
  Qt Network ............................. yes
  Qt Sql ................................. no
  Qt Testlib ............................. yes
  Qt Widgets ............................. yes
  Qt Xml ................................. no
Support enabled for:
  Using pkg-config ....................... yes
  QML debugging .......................... no
  udev ................................... no
  Using system zlib ...................... yes
Qt Core:
  DoubleConversion ....................... yes
    Using system DoubleConversion ........ no
  GLib ................................... no
  iconv .................................. no
  ICU .................................... no
  Logging backends:
    journald ............................. no
    syslog ............................... no
    slog2 ................................ no
  Using system PCRE2 ..................... no
Qt Network:
  getaddrinfo() .......................... no
  getifaddrs() ........................... yes
  IPv6 ifname ............................ yes
  libproxy ............................... no
  OpenSSL ................................ no
    Qt directly linked to OpenSSL ........ no
  SCTP ................................... no
  Use system proxies ..................... yes
Qt Gui:
  Accessibility .......................... yes
  FreeType ............................... yes
    Using system FreeType ................ no
  HarfBuzz ............................... yes
    Using system HarfBuzz ................ no
  Fontconfig ............................. no
  Image formats:
    GIF .................................. no
    ICO .................................. yes
    JPEG ................................. no
      Using system libjpeg ............... no
    PNG .................................. yes
      Using system libpng ................ yes
  EGL .................................... no
  OpenVG ................................. no
  OpenGL:
    Desktop OpenGL ....................... no
    OpenGL ES 2.0 ........................ no
    OpenGL ES 3.0 ........................ no
    OpenGL ES 3.1 ........................ no
  Session Management ..................... no
Features used by QPA backends:
  evdev .................................. yes
  libinput ............................... no
  INTEGRITY HID .......................... no
  mtdev .................................. no
  tslib .................................. no
  xkbcommon-evdev ........................ no
QPA backends:
  DirectFB ............................... no
  EGLFS .................................. no
  LinuxFB ................................ no
  VNC .................................... yes
  Mir client ............................. no
  X11:
    Using system-provided XCB libraries .. no
    EGL on X11 ........................... no
    Xinput2 .............................. no
    XCB XKB .............................. yes
    XLib ................................. yes
    XCB render ........................... yes
    XCB GLX .............................. yes
    XCB Xlib ............................. yes
    Using system-provided xkbcommon ...... no
Qt Widgets:
  GTK+ ................................... no
  Styles ................................. Fusion Windows
Qt PrintSupport:
  CUPS ................................... no
Qt Sql:
  DB2 (IBM) .............................. no
  InterBase .............................. no
  MySql .................................. no
  OCI (Oracle) ........................... no
  ODBC ................................... no
  PostgreSQL ............................. no
  SQLite2 ................................ no
  SQLite ................................. no
    Using system provided SQLite ......... no
  TDS (Sybase) ........................... no

Note: Using static linking will disable the use of dynamically
loaded plugins. Make sure to import all needed static plugins,
or compile needed modules into the library.

Note: -optimized-tools is not useful in -release mode.

Note: Disabling X11 Accessibility Bridge: D-Bus or AT-SPI is missing.

ERROR: Qt requires a compliant STL library.

ERROR: detected a std::atomic implementation that fails for function pointers.
Please apply the patch corresponding to your Standard Library vendor, found in
  qtbase/config.tests/atomicfptr

ERROR: Feature 'openssl-linked' was enabled, but the pre-condition '!features.securetransport && libs.openssl' failed.

ERROR: Feature 'openssl' was enabled, but the pre-condition '!features.securetransport && (features.openssl-linked || libs.openssl_headers)' failed.

ERROR: Feature 'system-freetype' was enabled, but the pre-condition 'features.freetype && libs.freetype' failed.

ERROR: Feature 'fontconfig' was enabled, but the pre-condition '!config.win32 && !config.darwin && features.system-freetype && libs.fontconfig' failed.
funcs.mk:243: recipe for target '/home/ubuntu/build/bitcoin/depends/work/build/powerpc64-linux-gnu/qt/5.9.7-72625d07ee6/qtbase/.stamp_configured' failed
make: *** [/home/ubuntu/build/bitcoin/depends/work/build/powerpc64-linux-gnu/qt/5.9.7-72625d07ee6/qtbase/.stamp_configured] Error 3
make: Leaving directory '/home/ubuntu/build/bitcoin/depends'

luke-jr commented:
Worked for me and the bot? :/

MarcoFalke commented on depends/README.md:
```diff
@@ -54,6 +56,10 @@ For linux AARCH64 cross compilation:
 
+For linux POWER 64-bit cross compilation (there are no packages for 32-bit):
+
+    sudo apt-get install curl g++-7-powerpc64-linux-gnu binutils-powerpc64-linux-gnu g++-7-powerpc64le-linux-gnu binutils-powerpc64le-linux-gnu
```
```suggestion
    sudo apt-get install g++-powerpc64-linux-gnu binutils-powerpc64-linux-gnu g++-powerpc64le-linux-gnu binutils-powerpc64le-linux-gnu
```

(Same as the other instructions)

MarcoFalke commented:
> Worked for me and the bot? :/

Oh, the failure is from the bot. But it passes for me locally on a fresh bionic vm:

```sh
apt update && apt install -y git vim htop make automake cmake curl g++-multilib libtool binutils-gold bsdmainutils pkg-config python3 && apt install -y  g++-powerpc64-linux-gnu binutils-powerpc64-linux-gnu g++-powerpc64le-linux-gnu binutils-powerpc64le-linux-gnu
git clone https://github.com/luke-jr/bitcoin --depth=1 --branch gitian_power64 && cd bitcoin/depends && make -j 9 HOST=powerpc64-linux-gnu qt

dongcarl commented:
Tested that Gitian building works on my own machine.

practicalswift commented:
Concept ACK

Diversity in testing is good. Thanks for working on this!

laanwj commented:
Would be nice to have this for 0.20.

laanwj commented:
> Also splits libpng out of Qt (since Qt's bundled copy is broken on POWER) and disables JPEG (since we don't use it).

Could this be solved by changing to a new version of Qt? If so, I'd prefer that.

MarcoFalke commented:
One of the commits has been cherry-picked and merged as 989fd539d5bf590c5f6070ee2a4a9e2d3018df2c

luke-jr commented on contrib/gitian-descriptors/gitian-linux.yml:
```diff
@@ -79,7 +85,11 @@ script: |
             echo "export FAKETIME=\"$1\"" >> ${WRAP_DIR}/${i}-${prog}
-            echo "exec \"\$REAL\" \"\$@\"" >> $WRAP_DIR/${i}-${prog}
+            if [ "${i:0:9}" = "powerpc64" ]; then
+                echo "exec \"\$REAL\" -mcpu=970 -mtune=power9 \"\$@\"" >> $WRAP_DIR/${i}-${prog}
```
I don't know how reliably those get passed everywhere. Also, you mean `CXXFLAGS`...

dongcarl replied:
:sweat_smile: yeah I meant CXXFLAGS...

with `depends` you can just modify `hosts/linux.mk`... I'm also not sure about tuning for a specific cpu here (perhaps I'm understanding incorrectly), especially because this will be activated for both big and small endian `HOST`s? Do you have sources for why these flags are needed?

luke-jr replied:
We don't want to force these CFLAGS on everyone... just use them for gitian builds so they perform well.

dongcarl replied:
Sure, then we can just supply `powerpc64_linux_CFLAGS='-mcpu=970 -mtune=power9' powerpc64le_linux_CFLAGS='-mcpu=970 -mtune=power9'`, but let's do that in a followup :-)

Wondering about the choice of `-mcpu` here... Why PowerPC 970 instead of a more generic `powerpc64` or `powerpc64le`? My reference: https://gcc.gnu.org/onlinedocs/gcc/RS_002f6000-and-PowerPC-Options.html#RS_002f6000-and-PowerPC-Options

fanquake replied:
> Wondering about the choice of -mcpu here... Why PowerPC 970 instead of a more generic powerpc64 or powerpc64le? 

This still atleast needs answering.

luke-jr replied:
I don't know. GCC doesn't let me compare `-mcpu` values AFAICT. Presumably generic options wouldn't be optimal for modern CPUs? (even 970 is pretty ancient)

luke-jr replied:
I'm told `-mcpu=970` is just powerpc64 + altivec.

laanwj commented:
This has needed rebase for a while.

hebasto commented:
@luke-jr 
> ...  Qt's bundled copy is broken on POWER...

Just for reference:
- [QTBUG-66388: libQtGui fails linking libqtpng on Power8+ and Power9](https://bugreports.qt.io/browse/QTBUG-66388)

sipa commented:
Would be nice to have this for 0.20.

luke-jr commented:
Rebased and fixed a bug (GCC 8 miscompiles for LE targets with `-mcpu=970`)

Considering the fragility of using the wrong `-mcpu`, I kept it in the wrapper...

MarcoFalke commented:
Excerpt:

```
+ make -j1 -C src check-security
make: Entering directory '/home/ubuntu/build/bitcoin/distsrc-powerpc64-linux-gnu/src'
Checking binary security...
bitcoind: failed NX
bitcoin-cli: failed NX
bitcoin-tx: failed NX
bitcoin-wallet: failed NX
test/test_bitcoin: failed NX
bench/bench_bitcoin: failed NX
qt/bitcoin-qt: failed NX
Makefile:18486: recipe for target 'check-security' failed
make: *** [check-security] Error 1
make: Leaving directory '/home/ubuntu/build/bitcoin/distsrc-powerpc64-linux-gnu/src'

luke-jr committed:
gitian: Always specify noexecstack since we enforce it unconditionally

luke-jr commented:
Fixed by making `noexecstack` explicit on all platforms (it doesn't make sense to rely on defaults when we require it...)

MarcoFalke commented:
Excerpt:

```
+ make -j1 -C src check-security
make: Entering directory '/home/ubuntu/build/bitcoin/distsrc-powerpc64-linux-gnu/src'
Checking binary security...
bitcoind: failed NX
bitcoin-cli: failed NX
bitcoin-tx: failed NX
bitcoin-wallet: failed NX
test/test_bitcoin: failed NX
bench/bench_bitcoin: failed NX
qt/bitcoin-qt: failed NX
Makefile:18794: recipe for target 'check-security' failed
make: *** [check-security] Error 1
make: Leaving directory '/home/ubuntu/build/bitcoin/distsrc-powerpc64-linux-gnu/src'

luke-jr commented:
Needs rebasing on top of #19751 once that's merged

luke-jr commented:
Rebased and re-tested.

laanwj commented on contrib/devtools/security-check.py:
```diff
@@ -145,7 +145,6 @@ def check_ELF_separate_code(executable):
     EXPECTED_FLAGS = {
         # Read + execute
         '.init': 'R E',
-        '.plt': 'R E',
```
Is `.plt` writable?
We should probably make this specific on architecture. If POWER9 is an exception here that needs the jump table to be writable (which would be kind of insecure in itself), let's make a specific exception for that.

fanquake replied:
Agree with @laanwj. No need to wholesale relax our checks just to accommodate POWER. A special case with a comment should be fine.

luke-jr replied:
Fixed

fanquake commented on contrib/gitian-descriptors/gitian-linux.yml:
```diff
@@ -38,13 +44,13 @@ script: |
   HOST_CFLAGS="-O2 -g"
   HOST_CXXFLAGS="-O2 -g"
-  HOST_LDFLAGS_BASE="-static-libstdc++ -Wl,-O2"
+  HOST_LDFLAGS="-static-libstdc++ -Wl,-O2 -Wl,-z,noexecstack"
```
If you're going to add `-z,noexecstack` in the gitian descriptor, it should have a comment explaining why it's here, rather than along with all of the other hardening options in configure, otherwise it's not going to be clear why this flag is special cased.

luke-jr replied:
I don't know the answer to that myself. This PR is merely making explicit what was already enforced by our scripts for other platforms.

luke-jr committed:
devtools/security-check: Abstract get_ELF_header out of check_ELF_PIE

luke-jr committed:
devtools/security-check: Allow .plt to be RW on PowerPC64

laanwj commented:
Now that #20434 is merged the symbol/security tooling-related changes should be able to be dropped here.

laanwj commented:
Would be good to get this in early in the 0.22 cycle. But it needs rebase; ping @luke-jr .

laanwj commented:
Closing in favor of #20963.
